sau/views.py

```diff
@@ -54,14 +55,40 @@ def index(request):
     sheep = get_all_sheep(request)
     return TemplateResponse(request, 'index.html', context={'sheep': sheep})
 
+def _get_sheep_or_404(request, slug):
+    if request.user.is_superuser:
+        return get_object_or_404(Sheep, slug=slug)
+
+    return get_object_or_404(Sheep, slug=slug,
+                             farm__farmers__in=[request.user])
 
 @login_required
 def sau(request, slug=""):
-    if request.user.is_superuser:
-        sheep = get_object_or_404(Sheep, slug=slug)
-    else:
-        sheep = get_object_or_404(Sheep, slug=slug,
-                                  farm__farmers__in=[request.user])
+    sheep = _get_sheep_or_404(request, slug)
     doses = Dose.get(sheep=sheep)
     return TemplateResponse(request, 'sau.html', context={'sheep': sheep,
                                                           'doses': doses})
+
+
+@login_required
+def _do_the_post(request, slug):
+    form = DoseForm(request.POST)
+    if form.is_valid():
```

Comment:
What happens if the form is not valid? I think Django is able to print error messages automatically (when using form.as_p), but only if you re-render the dose view with the same form.

In that case, the code would look like this:

    def dose(request, slug=''):
        form = None
        if request.method == "POST":
            form = DoseForm(request.POST)
            if form.is_valid()
                #  save and redirect
        elif request.method == "GET":
            form = DoseForm(initial=...)
        else:
            raise 404
        return TemplateResponse(..., ..., context={form: form, ...})

This way, you get an early redirect if everything is successful, otherwise you get to use the same TemplateResponse for both a "new" form and an old, erroneous one.